The new TimeSpan FromMilliseconds overload breaks System.Linq.Expressions

edited by @tarekgh to add the API proposal.

### Description

In .NET 9.0, we introduced overloads for the [`TimeSpan.From...`](https://github.com/dotnet/runtime/issues/93890#issuecomment-1777761585) methods to accept integral types. Unfortunately, the approved proposal included only a single overload with an optional parameter for the `FromMilliseconds` method.

```C#
public static TimeSpan FromMilliseconds(long milliseconds, long microseconds = 0);
```

Unfortunately, this broke the LINQ expression when used as follows, resulting in the error: `error CS0854: An expression tree may not contain a call or invocation that uses optional arguments`.

```C#
Expression<Action> a = () => TimeSpan.FromMilliseconds(1000); // compiles with .NET 8, fails with .NET 9
```

Note, this worked fine in net8 because we didn't have the new overload and we had the one that is accepting `double` parameter.


### Proposal

```diff
public partial struct TimeSpan
{
+    public static TimeSpan FromMilliseconds(long milliseconds);

+    public static TimeSpan FromMilliseconds(long milliseconds, long microseconds);
-    public static TimeSpan FromMilliseconds(long milliseconds, long microseconds = 0);
}
```

## Original Post


### Description

I have seen the breaking change notice with regards to F#, but migrating to .NET 9 has revealed a couple of places in our C# projects, where constructing a TimeSpan is part of a System.Linq.Expressions.Expression. These now don't compile with .NET 9


### Reproduction Steps

```csharp
using System.Linq.Expressions;
Expression<Action> a = () => TimeSpan.FromMilliseconds(1000); // compiles with .NET 8, fails with .NET 9
```


### Expected behavior

Code continues to compile

### Actual behavior

Code breaks in .NET 9

### Regression?

yes

### Known Workarounds

You need to force the `double` overload
```
Expression<Action> a = () => TimeSpan.FromMilliseconds(1000.0);
```
[or provide the optional parameter](https://github.com/dotnet/runtime/issues/109833#issuecomment-2477115423)
```
Expression<Action> a = () => TimeSpan.FromMilliseconds(1000, 0); 
```

### Configuration

_No response_

### Other information

The reason is that the optional parameters are not supported in expressions.

--- a/src/libraries/System.Private.CoreLib/src/System/TimeSpan.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/TimeSpan.cs
@@ -15,7 +15,7 @@ namespace System
     // to 100 nanoseconds. While this maps well into units of time such as hours
     // and days, any periods longer than that aren't representable in a nice fashion.
     // For instance, a month can be between 28 and 31 days, while a year
-    // can contain 365 or 366 days.  A decade can have between 1 and 3 leapyears,
+    // can contain 365 or 366 days.  A decade can have between 1 and 3 leap-years,
     // depending on when you map the TimeSpan into the calendar.  This is why
     // we do not provide Years() or Months().
     //
@@ -587,6 +587,18 @@ namespace System
 
         /// <summary>
         /// Initializes a new instance of the <see cref="TimeSpan"/> structure to a specified number of
+        /// milliseconds.
+        /// </summary>
+        /// <param name="milliseconds">Number of milliseconds.</param>
+        /// <returns>Returns a <see cref="TimeSpan"/> that represents a specified number of milliseconds.</returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// The parameter specify a <see cref="TimeSpan"/> value less than <see cref="MinValue"/> or greater than <see cref="MaxValue"/>
+        /// </exception>
+        public static TimeSpan FromMilliseconds(long milliseconds)
+            => FromUnits(milliseconds, TicksPerMillisecond, MinMilliseconds, MaxMilliseconds);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TimeSpan"/> structure to a specified number of
         /// milliseconds, and microseconds.
         /// </summary>
         /// <param name="milliseconds">Number of milliseconds.</param>
@@ -595,7 +607,7 @@ namespace System
         /// <exception cref="ArgumentOutOfRangeException">
         /// The parameters specify a <see cref="TimeSpan"/> value less than <see cref="MinValue"/> or greater than <see cref="MaxValue"/>
         /// </exception>
-        public static TimeSpan FromMilliseconds(long milliseconds, long microseconds = 0)
+        public static TimeSpan FromMilliseconds(long milliseconds, long microseconds)
         {
             Int128 totalMicroseconds = Math.BigMul(milliseconds, MicrosecondsPerMillisecond)
                                      + microseconds;

--- a/src/libraries/System.Runtime/ref/System.Runtime.cs
+++ b/src/libraries/System.Runtime/ref/System.Runtime.cs
@@ -6014,7 +6014,8 @@ namespace System
         public static System.TimeSpan FromMicroseconds(double value) { throw null; }
         public static System.TimeSpan FromMicroseconds(long microseconds) { throw null; }
         public static System.TimeSpan FromMilliseconds(double value) { throw null; }
-        public static System.TimeSpan FromMilliseconds(long milliseconds, long microseconds = (long)0) { throw null; }
+        public static System.TimeSpan FromMilliseconds(long milliseconds) { throw null; }
+        public static System.TimeSpan FromMilliseconds(long milliseconds, long microseconds) { throw null; }
         public static System.TimeSpan FromMinutes(double value) { throw null; }
         public static System.TimeSpan FromMinutes(long minutes) { throw null; }
         public static System.TimeSpan FromMinutes(long minutes, long seconds = (long)0, long milliseconds = (long)0, long microseconds = (long)0) { throw null; }

--- a/src/libraries/System.Runtime/tests/System.Runtime.Tests/System/TimeSpanTests.cs
+++ b/src/libraries/System.Runtime/tests/System.Runtime.Tests/System/TimeSpanTests.cs
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using System.Globalization;
+using System.Linq.Expressions;
 using System.Text;
 using Xunit;
 
@@ -631,7 +632,7 @@ namespace System.Tests
         [InlineData(0, -(maxSeconds + 1), 0, 0)]
         [InlineData(0, 0, maxMilliseconds + 1, 0)]
         [InlineData(0, 0, -(maxMilliseconds + 1), 0)]
-        [InlineData(0, 0, 0, maxMicroseconds + 1)]        
+        [InlineData(0, 0, 0, maxMicroseconds + 1)]
         [InlineData(0, 0, 0, -(maxMicroseconds + 1))]
         public static void FromMinutes_Int_ShouldOverflow(long minutes, long seconds, long milliseconds, long microseconds)
         {
@@ -713,7 +714,16 @@ namespace System.Tests
             long ticksFromMicroseconds = microseconds * TimeSpan.TicksPerMicrosecond;
             var expected = TimeSpan.FromTicks(ticksFromMilliseconds + ticksFromMicroseconds);
             Assert.Equal(expected, TimeSpan.FromMilliseconds(milliseconds, microseconds));
+
+            expected = TimeSpan.FromTicks(ticksFromMilliseconds);
+            Assert.Equal(expected, TimeSpan.FromMilliseconds(milliseconds));
+
+            // The following exist to ensure compilation of the expressions
+            Expression<Action> a = () => TimeSpan.FromMilliseconds(milliseconds);
+            Expression<Action> b = () => TimeSpan.FromMilliseconds(milliseconds, microseconds);
+            Expression<Action> c = () => TimeSpan.FromMilliseconds((double)milliseconds);
         }
+
         [Theory]
         [InlineData(maxMilliseconds + 1, 0)]
         [InlineData(-(maxMilliseconds + 1), 0)]
@@ -730,6 +740,11 @@ namespace System.Tests
         public static void FromMilliseconds_Int_ShouldOverflow(long milliseconds, long microseconds)
         {
             Assert.Throws<ArgumentOutOfRangeException>(() => TimeSpan.FromMilliseconds(milliseconds, microseconds));
+
+            if (microseconds == 0)
+            {
+                Assert.Throws<ArgumentOutOfRangeException>(() => TimeSpan.FromMilliseconds(milliseconds));
+            }
         }
 
         [Theory]

